woocommerce/woocommerce
# POC: Product traits compatibility

### Changes proposed in this Pull Request:

This POC adds a trait system for declaring compatibility with other traits.  It does the following:

* Sets up an abstraction where traits can be declared
* Moves trait specific methods into that class
* Adds trait metadata to blocks
* Hides blocks when incompatible traits are turned on

Follow-ups to consider for this PR:

* Allow compatibility for existing traits to be filtered (can technically already be done by using the `woocommerce_simple_product_traits` filter, but might be too complex)
* Replace and deprecate product types (POC in progress)
* Cross-trait compatibility (detecting declared or undeclared compatibility in the opposing trait)
* Server-side validation of trait compatibility



### How to test the changes in this Pull Request:
<img width="781" alt="Screenshot 2024-02-13 at 10 24 19 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/a4e18ca2-9171-4b43-81ce-f34101db909b">



1. Navigate to the Add New Product page with the new product editor experience enabled
2. Click on the "Virtual" toggle
3. Notice that the "Shipping" tab is hidden
4. Optionally make a call to `http://test.local/wp-json/wp/v2/template-parts?area=product-form` and notice the `productTrait` data in the block



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
